README.rst
=========== 
Sex Machine
===========

Installation::

    >>> pip install git+https://github.com/jreades/sexmachine.git

This package uses the underlying name frequency data from the program "gender" by Jorg Michael (described `here <http://www.autohotkey.com/community/viewtopic.php?t=22000>`_), and has been adapted from the original `Sex Machine <https://pypi.org/project/SexMachine/>`_ codebase to work with Python 3. Its use remains pretty straightforward::

    >>> import sexmachine.detector as gender
    >>> d = gender.Detector()
    >>> d.get_gender("Bob")
    'M'
    >>> d.get_gender("Sally")
    'F'
    >>> d.get_gender("Pauley") # should be androgynous
    'A'
    >>> d.get_gender("Tracey") # usually/mostly female
    'MF'

The result will be one of ``A`` (Androgynous), ``M`` (Male), ``F`` (Female), ``MM`` (Mostly Male), ``MF`` (Mostly Female) or ``U`` (Unknown). These values can be set to whatever you want using class-level variables as long as you do so before creating an actual detector::
    
    >>> import sexmachine.detector as gender
    >>> gender.Detector.male  = 'Male' # Class cars are: male, female, mmale, mfemale, unknown, androgynous
    >>> d = gender.Detector()
    >>> d.get_gender("Jon")
    'Male'

I18N is fully supported::

    >>> d.get_gender("Álfrún")
    'F'

Additionally, you can give preference to specific countries::

    >>> d.get_gender("Jamie")
    'MF'
    >>> d.get_gender("Jamie", 'Great Britain') # or 'great_britain' for backwards compatibility
    'MM'

Unlike the original SexMachine library this version uses a shared read-only dictionary so there is no particular penalty for creating multiple instances. Of course, if you want to change anything on the fly (e.g. a default return value) then you hae a very big problem.

As well, in the interests of keeping the package distribution as small as possible, the name_dict.txt file has been gzipped with concommitant changes to the code and reuqirements.

You can also now check the version::

    >>> import sexmachine
    >>> sexmachine.__version__
    '0.1.3'

Licenses
========

The generator code is distributed under the GPLv3.  The data file nam_dict.txt is released under the GNU Free Documentation License.


CHANGES.txt
v0.1.3, 09.02.2019 -- Use of class level variables to set string returned for each gender and replacement of ord() calculation with int(x,16)
v0.1.2, 08.02.2019 -- Update for Python3 based on isacarnekvist pull request.
v0.1.1, 19.04.2013 -- Data file.
v0.1.0, 17.04.2013 -- Initial release.


test.py
import os
import gzip
import operator

names = {}

gmappings = {
    'M':  'M',
    '1M': 'MM',
    '?M': 'MM',
    'F':  'F',
    '1F': 'MF',
    '?F': 'MF',
    '?':  'A',
    'NA': 'U'
}

mappings = ((256, ["<A/>"]),
            (257, ["<a/>"]),
            (258, ["<Â>"]),
            (259, ["<â>"]),
            (260, ["<A,>"]),
            (261, ["<a,>"]),
            (262, ["<C´>"]),
            (263, ["<c´>"]),
            (268, ["<C^>", "<CH>"]),
            (269, ["<c^>", "<ch>"]),
            (271, ["<d´>"]),
            (272, ["<Ð>", "<DJ>"]),
            (273, ["<ð>", "<dj>"]),
            (274, ["<E/>"]),
            (275, ["<e/>"]),
            (278, ["<E°>"]),
            (279, ["<e°>"]),
            (280, ["<E,>"]),
            (281, ["<e,>"]),
            (282, ["<Ê>"]),
            (283, ["<ê>"]),
            (287, ["<g^>"]),
            (290, ["<G,>"]),
            (286, ["<G^>"]),
            (291, ["<g´>"]),
            (298, ["<I/>"]),
            (299, ["<i/>"]),
            (304, ["<I°>"]),
            (305, ["<i>"]),
            (306, ["<IJ>"]),
            (307, ["<ij>"]),
            (310, ["<K,>"]),
            (311, ["<k,>"]),
            (315, ["<L,>"]),
            (316, ["<l,>"]),
            (317, ["<L´>"]),
            (318, ["<l´>"]),
            (321, ["<L/>"]),
            (322, ["<l/>"]),
            (325, ["<N,>"]),
            (326, ["<n,>"]),
            (327, ["<N^>"]),
            (328, ["<n^>"]),
            (336, ["<Ö>"]),
            (337, ["<ö>"]),
            (338, ["<OE>"]),
            (339, ["<oe>"]),
            (344, ["<R^>"]),
            (345, ["<r^>"]),
            (350, ["<S,>"]),
            (351, ["<s,>"]),
            (352, ["<S^>", "<SCH>", "<SH>"]),
            (353, ["<s^>", "<sch>", "<sh>"]),
            (354, ["<T,>"]),
            (355, ["<t,>"]),
            (357, ["<t´>"]),
            (362, ["<U/>"]),
            (363, ["<u/>"]),
            (366, ["<U°>"]),
            (367, ["<u°>"]),
            (370, ["<U,>"]),
            (371, ["<u,>"]),
            (379, ["<Z°>"]),
            (380, ["<z°>"]),
            (381, ["<Z^>"]),
            (382, ["<z^>"]),
            (7838, ["<ß>"]),
            )

def map_name(u):
    for code, patterns in mappings:
        for pattern in patterns:
            u = u.replace(pattern, chr(code))
    return u

COUNTRIES = [ x.strip() for x in """Great Britain, Ireland, USA, Italy, Malta, Portugal, Spain, France, 
                   Belgium, Luxembourg, The Netherlands, East Frisia, Germany, Austria, 
                   Switzerland, Iceland, Denmark, Norway, Sweden, Finland, Estonia, Latvia, 
                   Lithuania, Poland, Czech Republic, Slovakia, Hungary, Romania, 
                   Bulgaria, Bosnia and Croatia, Kosovo, Macedonia, Montenegro, Serbia, 
                   Slovenia, Albania, Greece, Russia, Belarus, Moldova, Ukraine, Armenia, 
                   Azerbaijan, Georgia, The Stans, Turkey, Arabia, Israel, China, India, 
                   Japan, Korea, Vietnam, Other
                 """.split(",") ]
countries = {k: v for v, k in enumerate(COUNTRIES)}
for v, k in enumerate(COUNTRIES):
    countries[k.lower().replace(" ","_")] = v

def _parse(filename):
    """Opens data file and for each line, calls _eat_name_line"""
    with gzip.open(filename, 'r') as f:
        for line in f:
            _eat_name_line(line.decode('iso8859-1').strip())

def _eat_name_line(line):
    """Parses one line of data file"""
    if line[0] not in "#=":
        gender = line[0:2].strip()
        name   = map_name(line[3:29].strip())
        freq   = line[30:-1].replace(" ","0")

        _set(name, gmappings[gender], freq)

def _set(name, gender, country_values):
    """Sets gender and relevant country values for names dictionary of detector"""
    if '+' in name:
        for replacement in ['', ' ', '-']:
            _set(name.replace('+', replacement), gender, country_values)
    else:
        if name not in names:
            names[name] = {}
        names[name][gender] = country_values

def dump_name(country_values):
    for key, val in country_values.items(): 
        print(key)
        for i in range(0, len(COUNTRIES)):
            if val[i] != '0': print("\t" + COUNTRIES[i] + " -> " + val[i] + " -> " + str(int(val[i],16)))

# def get_gender(name, country=None):
#         """Returns best gender for the given name and country pair"""
#         if name not in names:
#             return Null
#         elif not country:
#             def counter(country_values):
#                 print(",".join(country_values))
#                 country_values = country_values.replace(" ", "")
#                 #print(sum(list(map(lambda c: int(c) > 64 and int(c)-55 or int(c)-48, country_values))))
#                 return (len(list(country_values)),
#                         sum(list(map(lambda c: int(c) > 64 and int(c)-55 or int(c)-48, country_values))))
#             return self._most_popular_gender(name, counter)
#         elif country in self.__class__.COUNTRIES:
#             index = self.__class__.COUNTRIES.index(country)
#             print("Index: " + str(index))
#             counter = lambda e: (try_int(e[index])-32, 0)
#             return self._most_popular_gender(name, counter)
#         else:
#             raise NoCountryError("No such country: %s" % country)

_parse(os.path.join(os.path.dirname(__file__), "sexmachine/data/nam_dict.txt.gz"))

#print(names['Yun Cha'])
#print(names['Yun-Cha'])
#print(names['Maria'])
print(names['Jamie'])

#dump_name(names['Maria'])
#print(" ")

def name_freq(country_values):
    return sum(list(map(lambda c: int(c,16), country_values)))

# Needs a tie-breaker option based on default values? Or a 'U' value
# for 'we don't know'? Something different from Androgyne? Guess this is 
# where the number of countries comes into it, though seems to me you
# also want to weight by country population! Maybe a strict=T/F param
# based on whether you want it to infer a value or be specific to that 
# country?
def country_prob(lkp, ctry, strict=False):
    
    ctry_results = {} # Store the by-country results
    glob_results = {} # Store the global results 

    ix = COUNTRIES.index(ctry)
    for key, val in lkp.items():
        ctry_results[key] = int(val[ix],16)
        glob_results[key] = name_freq(val)
    
    if max(ctry_results.values()) == 0: 
        if strict is False:
            return max(glob_results.items(), key=operator.itemgetter(1))[0]
        else:
            return gmappings['NA']
    else:
         return max(ctry_results.items(), key=operator.itemgetter(1))[0]

print("Jamie")
dump_name(names['Jamie'])
print(country_prob(names['Jamie'],'Great Britain',strict=True))

print("Maria")
dump_name(names['Maria'])
print(country_prob(names['Maria'],'Italy',strict=False))

sexmachine/test/test_mapping.py
# -*- coding: utf-8 -*-
import unittest
import sexmachine.mapping as m

class TestMapping(unittest.TestCase):

    def test_map_name(self):
        self.assertEqual(m.map_name(u"Ay<s,>e"), u"Ayşe")
        self.assertEqual(m.map_name(u"<ß>ahri"), u"ẞahri")
        self.assertEqual(m.map_name(u"<Ö>mer"), u"Őmer")
        self.assertEqual(m.map_name(u"<SCH>vet"), u"Švet")

if __name__ == '__main__':
    unittest.main()


sexmachine/test/test_detector.py
# -*- coding: utf-8 -*-
import unittest
import sexmachine.detector as d

class TestDetector(unittest.TestCase):

    def setUp(self):
        self.case = d.Detector()

    def test_gender(self):
        self.assertEqual(self.case.get_gender("Bob"), "M")
        self.assertEqual(self.case.get_gender("Sally"), "F")
        self.assertEqual(self.case.get_gender("Pauley"), "A")

    def test_unicode(self):
        self.assertEqual(self.case.get_gender("Álfrún"), "F")
        self.assertEqual(self.case.get_gender("Ayşe"), "F")
        self.assertEqual(self.case.get_gender("Gavriliţă"), "F")
        self.assertEqual(self.case.get_gender("İsmet"), "M")
        self.assertEqual(self.case.get_gender("Snæbjörn"), "M")

    def test_country(self):
        self.assertEqual(self.case.get_gender("Jamie"), "MF")
        self.assertEqual(self.case.get_gender("Jamie", "great_britain"), "MM")
        self.assertEqual(self.case.get_gender("Jamie", "Great Britain"), "MM")

if __name__ == '__main__':
    unittest.main()


sexmachine/detector.py
import os
import gzip
import operator
from .mapping import map_name

class NoCountryError(Exception):
    """Raised when unknown country is queried"""
    pass

class Detector(object):
    """Get gender by first name"""

    names     = {}
    countries = {}

    default_fn = os.path.join('data','nam_dict.txt.gz')

    male        = 'M'
    female      = 'F'
    androgynous = 'A'
    unknown     = 'U'
    mmale       = 'MM'
    mfemale     = 'MF'

    COUNTRIES = [ x.strip() for x in """Great Britain, Ireland, USA, Italy, Malta, Portugal, Spain, France, 
                   Belgium, Luxembourg, The Netherlands, East Frisia, Germany, Austria, 
                   Switzerland, Iceland, Denmark, Norway, Sweden, Finland, Estonia, Latvia, 
                   Lithuania, Poland, Czech Republic, Slovakia, Hungary, Romania, 
                   Bulgaria, Bosnia and Croatia, Kosovo, Macedonia, Montenegro, Serbia, 
                   Slovenia, Albania, Greece, Russia, Belarus, Moldova, Ukraine, Armenia, 
                   Azerbaijan, Georgia, The Stans, Turkey, Arabia, Israel, China, India, 
                   Japan, Korea, Vietnam, Other
                 """.split(",") ]

    def __init__(self):
        """Create a new Gender Detector object

        Args:

        Returns:
            :obj: Object ot type Detector
        """

        # If the countries dictionary hasn't been loaded
        # yet then we will load it here -- this is shared
        # between all objects of this class and so should
        # be read-only
        if len(Detector.countries) == 0:
            Detector.countries = {k: v for v, k in enumerate(Detector.COUNTRIES)}
            for v, k in enumerate(Detector.COUNTRIES):
                Detector.countries[k.lower().replace(" ","_")] = v

        # This maps the inputs from the nam_dict.txt
        # file in to a standard gender descriptor --
        # these can be modified fairly easily prior
        # to instantiation of the first object of this
        # class. After that the results may be fairly
        # unpredictable: there is a reload method that
        # may enable this to be changed afterwards but
        # I have not properly tested it.
        self.mapping = {
            'M':  Detector.male,
            'F':  Detector.female,
            '1M': Detector.mmale,
            '?M': Detector.mmale,
            '1F': Detector.mfemale,
            '?F': Detector.mfemale,
            '?':  Detector.androgynous
        }
        
        # Load the nam_dict data file.
        self.load()

    def _parse(self, filename):
        """Private method: opens data file and for each line, calls _eat_name_line"""
        with gzip.open(filename, 'r') as f:
            for line in f:
                self._eat_name_line(line.decode('iso8859-1').strip())

    def _eat_name_line(self, line):
        """Parses one line of data file and converts
        it into a data representation.
        
        Lines starting '#' or '=' are skipped.
        
        The rest of the line is handled as follows:
        - Characters 0-2 are interpreted as gender (M, F, 1M, 1F, ?M, ?F, ?)
        - Characters 3-29 are interpreted as the name
        - Characters 30-End are interpreted as data (each column/character is a hexadecimal number for a country)
        """
        if line[0] not in "#=":
            gender = line[0:2].strip()
            name   = map_name(line[3:29].strip())
            freq   = line[30:-1].replace(" ","0")

            self._set(name, gender, freq)

    def _set(self, name, gender, country_values):
        """Sets gender and relevant country values into the data structure."""
        
        # Sets variations on spelling/hyphenation
        if '+' in name:
            for replacement in ['', ' ', '-']:
                self._set(name.replace('+', replacement), gender, country_values)
        else:
            if name not in Detector.names:
                Detector.names[name] = {}

            Detector.names[name][self.mapping[gender]] = country_values
    
    def dump_name(self, name):
        """Useful debugging/inspection tool for the data structure.

        Args:
            name (str): The name to inspect.

        Returns:
            None: No return value, it simply outputs information for each country.

        """
        if len(Detector.names) == 0:
            self._parse(os.path.join(os.path.dirname(__file__), Detector.default_fn))

        for key, val in Detector.names[name].items(): 
            print(key)
            for i in range(0, len(Detector.COUNTRIES)):
                if val[i] != '0': print("\t" + Detector.COUNTRIES[i] + " -> " + str(int(val[i],16)))
    
    def _name_freq(self, country_values):
        """Calculates a sum across all hexadecimal country values for a given name."""
        return sum(list(map(lambda c: int(c,16), country_values)))

    def _max_prob(self, ds):
        """Works out which gender is 'most likely' (max probability)
        based on a data structure that contains either the sum for
        each gender or the value for a specific country."""
        mv = max(ds.items(), key=operator.itemgetter(1))[0]
        
        if ds[mv] == 0:
            return Detector.unknown
        else:
            return mv

    def _global_prob(self, name, strict=False):
        """Private: works out a likelihood for a given name.

        Args:
            name (str): The name to search for.
            strict (bool): Should search only on specified country. Ignored by this method but ensures consistent interface.

        Returns:
            str: The gender thought most likely to be associated with the name.

        """
        glob_results = {} # Store the global results 

        for key, val in Detector.names[name].items():
            glob_results[key] = self._name_freq(val)

        return self._max_prob(glob_results)

    def _country_prob(self, name, ctry, strict=False):
        """Private: works out a likelihood for a given name.

        I'm unclear as to whether these results should be weighted by the
        population of the countries (e.g. a '1' for USA and a '1' for 
        Luxembourg should not be treated equally when trying to work out
        the likely gender).
        
        The strict parameter set to True means that _only_ the specified 
        country is considered when inferring gender. If false (the default)
        then _if_ no specific country value is found then the global 
        distribution will be used. If true (the other option) then it 
        will return _only_ the country value _even if_ that value is 0.

        Args:
            name (str): The name to search for.
            strict (bool): Should search be confined to only the specified country. Defaults to False.

        Returns:
            str: The gender thought most likely to be associated with the name.

        """
        
        ctry_results = {} # Store the by-country results
        glob_results = {} # Store the global results 

        ix = self.countries[ctry]
        for key, val in Detector.names[name].items():
            ctry_results[key] = int(val[ix],16)
            glob_results[key] = self._name_freq(val)
        
        if strict is False and max(ctry_results.values()) == 0: 
            return self._max_prob(glob_results)
        else:
            return self._max_prob(ctry_results)

    def load(self):
        """Used to load the data. Normally called by __init__."""
        if len(Detector.names) == 0:
            self._parse(os.path.join(os.path.dirname(__file__), Detector.default_fn))
    
    def reload(self):
        """May enable the data set to be reloaded with a new set of return values based on the class variables."""
        Detector.names = {}
        self.load()

    def get_gender(self, name, country=None, strict=False):
        """Returns best gender for the given name and country pair.
        
        The strict parameter set to True means that _only_ the specified
        country is considered when inferring gender. If false (the default)
        then _if_ no specific country value is found then the global
        distribution will be used. If true (the other option) then it
        will return _only_ the country value _even if_ that value is 0.

        Args:
            name (str): The name to search for.
            ctry (str): The name of the country. Defaults to None for a global search.
            strict (bool): Should search be confined to only the specified country. Defaults to False.

        Returns:
            str: The gender thought most likely to be associated with the name.
        """
        if len(Detector.names) == 0:
            self._parse(os.path.join(os.path.dirname(__file__), Detector.default_fn))

        if name not in Detector.names:
            return Detector.unknown

        elif not country:
            return self._global_prob(name)

        elif country in Detector.countries:
            return self._country_prob(name, country, strict)

        else:
            raise NoCountryError("No such country: %s" % country)
